### PR TITLE
feat(ci): self-heal workflow — analyze test failures and propose fixes (Phase 1)

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

- Adds `.github/workflows/self-heal.yml` — triggers on `workflow_run` completion when the Test workflow fails on a PR
- Adds `scripts/github_ci_self_heal.py` — calls Claude (Haiku) with the failure log + PR diff and returns a root-cause/fix-proposal analysis
- Posts the analysis as a PR comment (comment-only, no auto-merge or auto-push)

## Why `workflow_run` instead of `pull_request`

`workflow_run` runs in the context of the base repo, so `ANTHROPIC_API_KEY` is available even for fork PRs — unlike `pull_request` which blocks secrets from forks for security.

## Flow

```
Test workflow fails on PR
  → self-heal triggers (workflow_run)
  → downloads failure log (gh run view --log-failed)
  → downloads PR diff (gh pr diff)
  → calls Claude Haiku via scripts/github_ci_self_heal.py
  → posts analysis comment on the PR
```

## Example comment output

> **Root cause**: The new `parse_foo()` function doesn't handle empty string input, causing `test_parse_empty` to fail with `AttributeError`.
>
> **Proposed fix**: Add a guard at the top of `parse_foo()`: `if not s: return None`
>
> **Confidence**: high — the diff adds the function and the test directly tests the edge case.

## Non-goals (Phase 1)

- Auto-merge of fixes
- Creating fix PRs automatically
- Cross-repo propagation

## Test plan

- [ ] CI passes on this PR
- [ ] Manually verify by checking workflow triggers with a deliberately broken test on a test PR

### CODE REVIEW

File: .github/workflows/self-heal.yml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            echo ""
+            cat /tmp/analysis.txt
+            echo ""
+            echo "---"
+            echo "*[Failing run](${FAILED_RUN_URL}) · [Analysis run](${BOT_RUN_URL})*"
+          } > /tmp/comment.md
+
+          gh pr comment ${{ steps.pr.outputs.number }} \
+            --repo ${{ github.repository }} \
+            --body-file /tmp/comment.md
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply .github/workflows/self-heal.yml:138]: Addressed in bf93732a8: added a `<!-- gptme-self-heal -->` HTML marker to posted comments and a dedup guard that checks for that marker before posting. Repeated CI failures on the same PR will skip posting if a comment already exists.

File: .github/workflows/self-heal.yml [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.workflow_run.head_sha }}
+
+      - name: Get PR number
+        id: pr
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          PR_NUMBER="${{ github.event.workflow_run.pull_requests[0].number }}"
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply .github/workflows/self-heal.yml]: Fixed in bf93732a8: replaced `pull_requests[0].number` with a SHA-based lookup via the REST API (`/pulls?state=open&per_page=100` → `select(.head.sha == HEAD_SHA)`). Fork PRs are now handled correctly.

File: scripts/github_ci_self_heal.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    pr_diff_path = Path(sys.argv[2])
+
+    failure_log = (
+        failure_log_path.read_text()
+        if failure_log_path.exists()
+        else "(no failure log available)"
+    )
+    pr_diff = (
+        pr_diff_path.read_text() if pr_diff_path.exists() else "(no diff available)"
+    )
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply scripts/github_ci_self_heal.py:78]: Fixed in bf93732a8: added `errors='replace'` to both `read_text()` calls so ANSI/non-UTF-8 bytes in log output are replaced rather than raising `UnicodeDecodeError`.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: .github/workflows/self-heal.yml [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        run: |
+          # pull_requests[] is empty for fork PRs — look up by head SHA instead.
+          HEAD_SHA="${{ github.event.workflow_run.head_sha }}"
+          PR_NUMBER=$(gh api "repos/${{ github.repository }}/pulls?state=open&per_page=100" \
+            --jq ".[] | select(.head.sha == \"$HEAD_SHA\") | .number" | head -1)
+          if [ -z "$PR_NUMBER" ]; then
+            echo "No open PR found for sha $HEAD_SHA — skipping"
+            exit 0
+          fi
+          echo "number=$PR_NUMBER" >> "$GITHUB_OUTPUT"
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply .github/workflows/self-heal.yml:60]: Fixed in fd246f60c: replaced `exit 0` with a `found=true/false` output and added `if: steps.pr.outputs.found == 'true'` to every downstream step. When no open PR matches the triggering SHA the job now short-circuits cleanly — no downstream steps run, no invalid URLs hit.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: scripts/github_ci_self_heal.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    if not failure_log_path.exists():
+        print(f"Warning: failure log not found at {failure_log_path}", file=sys.stderr)
+    if not pr_diff_path.exists():
+        print(f"Warning: PR diff not found at {pr_diff_path}", file=sys.stderr)
+
+    analysis = analyze_failure(failure_log, pr_diff)
+    print(analysis)
+    return 0
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply scripts/github_ci_self_heal.py:90]: Fixed in cddd262b2: added an empty-response guard in `main()` — if `analyze_failure()` returns an empty string the script now prints to stderr and returns exit code 1, which causes the `Analyze failure` step to fail and the error-fallback comment to fire instead of posting an empty analysis.

File: .github/workflows/self-heal.yml [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        if: steps.pr.outputs.found == 'true' && success() && steps.analyze.outcome == 'success'
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          FAILED_RUN_URL: ${{ github.event.workflow_run.html_url }}
+          BOT_RUN_URL: ${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}
+          PR_NUMBER: ${{ steps.pr.outputs.number }}
+        run: |
+          # Dedup: skip if a self-heal comment already exists for this PR.
+          EXISTING=$(gh api "repos/${{ github.repository }}/issues/${PR_NUMBER}/comments" \
+            --jq '[.[] | select(.body | contains("<!-- gptme-self-heal -->"))] | length')
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply .github/workflows/self-heal.yml]: Fixed in cddd262b2: added `--paginate` to the dedup `gh api` call and switched from `[.[] | ...] | length` (broken with paginate) to `.[] | ... | .id | wc -l` which counts correctly across all pages.

File: .github/workflows/self-heal.yml [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+          {
+            echo "<!-- gptme-self-heal -->"
+            echo "## 🔧 gptme-bot: Self-heal (error)"
+            echo ""
+            echo "I attempted to analyze the CI failure but encountered an error."
+            echo "[Check the bot run](${BOT_RUN_URL}) for details."
+          } > /tmp/error-comment.md
+          gh pr comment "${PR_NUMBER}" \
+            --repo ${{ github.repository }} \
+            --body-file /tmp/error-comment.md
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply .github/workflows/self-heal.yml:164]: Fixed in cddd262b2: added the same `EXISTING` dedup guard (with `--paginate`) to the error-fallback step. Both paths now skip posting if a `<!-- gptme-self-heal -->` comment already exists.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: .github/workflows/self-heal.yml [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+      github.event.workflow_run.conclusion == 'failure' &&
+      github.event.workflow_run.event == 'pull_request'
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+
+    steps:
+      - name: Checkout head commit
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.workflow_run.head_sha }}
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply .github/workflows/self-heal.yml]: Fixed in e9c0bbc: removed `ref: ${{ github.event.workflow_run.head_sha }}` from the checkout step. The checkout now defaults to the base branch (master), so `scripts/github_ci_self_heal.py` always runs from the trusted base repo — not the fork's code. No fork PR author can modify the script to exfiltrate `ANTHROPIC_API_KEY` or the `pull-requests: write` token.

File: .github/workflows/self-heal.yml [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+      - name: Download failure log
+        if: steps.pr.outputs.found == 'true'
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh run view ${{ github.event.workflow_run.id }} \
+            --repo ${{ github.repository }} \
+            --log-failed 2>&1 | tail -c 15000 > /tmp/failure.log
+          echo "Failure log size: $(wc -c < /tmp/failure.log) bytes"
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply .github/workflows/self-heal.yml:72]: Fixed in b53ec79: added `set -o pipefail` to the Download failure log step. Now if `gh run view` exits non-zero (transient API error, rate limit, etc.), the pipeline propagates that exit code and the step fails rather than silently continuing with an empty or garbage log file.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: .github/workflows/self-heal.yml [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+            echo "---"
+            echo "*[Failing run](${FAILED_RUN_URL}) · [Analysis run](${BOT_RUN_URL})*"
+          } > /tmp/comment.md
+
+          gh pr comment "${PR_NUMBER}" \
+            --repo ${{ github.repository }} \
+            --body-file /tmp/comment.md
+
+      - name: Report error if analysis failed
+        if: steps.pr.outputs.found == 'true' && failure() && (steps.analyze.outcome == 'failure' || steps.post_comment.outcome == 'failure')
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply .github/workflows/self-heal.yml]: Fixed in 111f4df: simplified the condition to `steps.pr.outputs.found == 'true' && failure()`. When `Download failure log` fails via pipefail, downstream steps get `outcome=='skipped'` (not `'failure'`) because they run under the default `success()` guard — the old condition never matched. The dedup guard inside the step still prevents double-posting.

**Comment by TimeToBuildBob** [comment]
@greptileai review
